### DISCUSSION TITLE
Leaks and High memory usage

### DISCUSSION THREAD

**Discussion by dirtyhabits97** [MobileNativeFoundation/discussions#117] [General]
I’m curious about how other engineers measure memory leaks / high memory usage (manually or automatically) during development and/or production.

**Comment by dirtyhabits97** [comment]
At Rappi we currently measure it manually using instruments and the memory-graph tool in Xcode, but we want to automate this process and generate a weekly report to notify teams with the most leaks and highest memory usage in given screens. I have 2 approaches in mind (both involving UI tests):

1. Run UITests with `xctrace` attached in order to get a `.trace` file which we can then query to get the needed information. I saw a similar implementation [here](https://tech.showmax.com/2019/02/automated-mem-leak-testing-ios/), however `instruments` and `TraceUtility` are both deprecated in favor of `xctrace`.

2. Run UITests with `XCTMemoryMetric`  in order to log memory usage. However, this approach would give us less information about the leaks. As per the [release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-13-beta-release-notes), it looks like Xcode 13 will output a series of `memgraphs`  when using `XCTMemoryMetrics` which would help us when generating a report.

Have you guys done something similar?

**Comment by imagiczhang** [comment]
Besides memory, is there any other app health metrics we collect in mobile devlopment? e.g. CPU, threads pool, scroll, I/O, battery usage and etc?

**Comment by kazimdsaidul** [comment]
In android a good third party library you can use. 

https://square.github.io/leakcanary/
